### ISSUE TITLE
Mails do not get sent to users when exim is used as mail daemon (sendmail with -t command).

### ISSUE THREAD

**Issue by pspoerri** [gitlabhq/gitlabhq#4866]
1. **Summary:** Mails do not get sent to users when exim is used as mail daemon.
2. **Steps to reproduce:** Setup gitlab (5.2 and 6.0) on linux with exim as sendmail client. 
3. **Expected behavior:** Mail should arrive at users location.
4. **Observed behavior:** Gitlab will notify that a mail to $user has been sent but exim won't report that it sent a mail to $user. Exim will also report 

    ``` 
From: Mail Delivery System <Mailer-Daemon@example.com>
Subject: Mail failure - no recipient addresses
Date: 29. Juli 2013 13:51:23 MESZ
To: <ops@example.com>
A message that you sent using the -t command line option contained no
addresses that were not also on the command line, and were therefore
suppressed. This left no recipient addresses, and so no delivery could
be attempted.
```

6. **Output of checks**
    * Results of GitLab [Application Check]

      ```
# bundle exec rake gitlab:check RAILS_ENV=production
Checking Environment ...
Git configured for git user? ... yes
Has python2? ... yes
python2 is supported version? ... yes
Checking Environment ... Finished
Checking GitLab Shell ...
GitLab Shell version >= 1.7.0 ? ... OK (1.7.0)
Repo base directory exists? ... yes
Repo base directory is a symlink? ... no
Repo base owned by git:git? ... yes
Repo base access is drwxrws---? ... yes
post-receive hook up-to-date? ... yes
post-receive hooks in repos are links: ...
Checking GitLab Shell ... Finished
Checking Sidekiq ...
Running? ... yes
Checking Sidekiq ... Finished
Checking GitLab ...
Database config exists? ... yes
Database is SQLite ... no
All migrations up? ... yes
GitLab config exists? ... yes
GitLab config outdated? ... no
Log directory writable? ... yes
Tmp directory writable? ... yes
Init script exists? ... no
  Try fixing it:
  Install the init script
  For more information see:
  doc/install/installation.md in section "Install Init Script"
  Please fix the error above and rerun the checks.
Init script up-to-date? ... can't check because of previous errors
Projects have satellites? ...
[...]
Redis version >= 2.0.0? ... yes
Your git bin path is "/usr/bin/git"
Git version >= 1.7.10 ? ... no
  Try fixing it:
  Update your git to a version >= 1.7.10 from 1.7.9
  Please fix the error above and rerun the checks.
      ```
      I know that git has not been updated yet. But it doesn't have anything to do with gitlab email. Hence I'm leaving it there.

    * Sha1: 
      4c8f3699f1f1334f1b7ed964d4b6ca985796211a

7. **Possible fixes**:

```
        diff --git a/config/application.rb b/config/application.rb
        index f7349fa..75eee2c 100644
        --- a/config/application.rb
        +++ b/config/application.rb
        @@ -11,6 +11,8 @@ end
        
         module Gitlab
           class Application < Rails::Application
        +    config.action_mailer.sendmail_settings = { :arguments => "-i" }
        +
             # Settings in config/environments/* take precedence over those specified he
             # Application configuration should go into files in config/initializers
             # -- all .rb files in that directory are automatically loaded.
```

This issue was also discussed in #3943. I thought to reopen it but since it was closed by the requestor. I thought it might be better to just create a new issue.

The issue is also discussed here: https://github.com/mikel/mail/issues/70

**Comment by yoe** [comment]
The exim 4 info documentation has this to say about -t:

     If the command has any arguments, they specify addresses to which
     the message is _not_ to be delivered. That is, the argument
     addresses are removed from the recipients list obtained from the
     headers. This is compatible with Smail 3 and in accordance with
     the documented behaviour of several versions of Sendmail, as
     described in man pages on a number of operating systems (e.g.
     Solaris 8, IRIX 6.5, HP-UX 11). However, some versions of Sendmail
     _add_ argument addresses to those obtained from the headers, and
     the O'Reilly Sendmail book documents it that way. Exim can be made
     to add argument addresses instead of subtracting them by setting
     the option `extract_addresses_remove_arguments' false.

IOW, it's not a bug in exim, it's just a matter of badly specified "standards", and conflicting defaults. The fix is to, simply, add a line `extract_addresses_remove_arguments = false` to the global section of exim's configuration file...

**Comment by pravi** [comment]
Can this workaround be mentioned in the installation document? Switching to postfix for just one line configuration change seems not worth it.

**Comment by dosire** [comment]
@pravi Feel free to suggest a non-intrusive merge request to change the manual installation doc.

**Comment by isti37** [comment]
_extract_addresses_remove_arguments = false_ in /etc/exim/exim.conf saved the day because I couldn't get smtp either to work with gitlab.

**Comment by emmenlau** [comment]
On Debian and Ubuntu, with a *split* exim4 config, the following worked for me:
Create a new file /etc/exim4/conf.d/main/10_exim4-gitlab_config_options with contents
````
# GitLab workaround against a problem how exim4 handles the -t option:
extract_addresses_remove_arguments=false
````
Then regenerate configs and restart exim4 with:
````
sudo update-exim4.conf
sudo service exim4 restart
````
If you already have frozen mails that could not be delivered before, you can send the whole queue with
````
sudo exim4 -qff
````

**Comment by yoe** [comment]
I'd just like to add that the bug really is in gitlab for using a command-line option that is not necessary and has wildly different results for different MTA's. There are plenty of MUA's which support sending mail with sendmail that this shouldn't be an issue.

The proper fix, really, is to either drop use of the `-t` command-line option and to specify *all* recipient addresses to which mail should be sent on the command line, *or* to retain it but then to specify *all* recipient addresses in the mail rather than on the command line.

Anything else is not portable and should be avoided.

**Comment by chewi** [comment]
I just ran into this myself so yes, it is annoying, but let's be clear, it is a Rails issue. It was also an issue in mikel/mail#70 but that's since been fixed. Even in current master, Rails overrides the arguments used by Mail [here](https://github.com/rails/rails/blob/7f18ea14c893cb5c9f04d4fda9661126758332b5/actionmailer/lib/action_mailer/delivery_methods.rb#L39). I think rails/rails#1755 should be reopened.

**Comment by jvanbaarsen** [comment]
We are currently in the process of closing down the issue tracker on GitHub, to prevent duplication with the GitLab.com issue tracker. Since this is an older issue I'll be closing this for now. If you think this is still an issue I encourage you to open it on the [GitLab.com issue tracker](https://gitlab.com/gitlab-org/gitlab-ce/issues).

**Comment by shane-axiom** [comment]
Still relevant, migrated to GitLab's issue tracker here: https://gitlab.com/gitlab-org/gitlab-ce/issues/12754

**Comment by dosire** [comment]
Thanks for moving @shane-axiom

**Comment by chewi** [comment]
FYI this is fixed in Rails master and the 4.2 branch so it should hit 4.2.7.
